File: retrocast-rs/crates/retrocast-core/src/adapters/synllama.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{Adapter, RawRouteEntry, common};
use crate::{
    chem,
    error::Result,
    model::{Route, Target},
    route::AdaptMode,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
struct SynLlamaRoute {
    synthesis_string: String,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SynLlamaAdapter;

impl Adapter for SynLlamaAdapter {
    fn name(&self) -> &'static str {
        "synllama"
    }

    fn entries(&self, payload: Value, source_key: Option<&str>) -> Result<Vec<RawRouteEntry>> {
        let routes: Vec<SynLlamaRoute> = serde_json::from_value(payload)
            .map_err(|error| common::schema(self.name(), format!("invalid route list: {error}")))?;
        Ok(routes
            .into_iter()
            .enumerate()
            .map(|(index, route)| RawRouteEntry {
                payload: serde_json::to_value(route).expect("SynLlama route is serializable"),
                source_key: source_key.map(str::to_owned),
                source_row_index: None,
                source_record_id: None,
                target_hint_id: None,
                target_hint_smiles: None,
                source_order: Some(index + 1),
            })
            .collect())
    }

    fn cast(&self, raw_route: Value, mode: AdaptMode, target: Option<&Target>) -> Result<Route> {
        let route: SynLlamaRoute = serde_json::from_value(raw_route)
            .map_err(|error| common::schema(self.name(), format!("invalid route: {error}")))?;
        let parts: Vec<_> = route
            .synthesis_string
            .split(';')
            .map(str::trim)
            .filter(|part| !part.is_empty())
            .collect();
        if parts.is_empty() {
            return Err(common::logic(
                self.name(),
                "adapter.route_string_empty",
                "empty synthesis string",
            ));
        }
        let parsed_target = match chem::normalize(parts.last().expect("parts is nonempty")) {
            Ok((smiles, _)) => smiles,
            Err(_) if mode == AdaptMode::Prune => {
                return Err(common::target_pruned(self.name()));
            }
            Err(error) => return Err(error),
        };
        if let Some(target) = target {
            common::require_target_match(&parsed_target, &target.smiles, &target.id, self.name())?;
        }
        let precursor_map = parse_synthesis(&parts, mode, self.name())?;
        let route_target = common::build_from_precursor_map(
            &parsed_target,
            &precursor_map,
            self.name(),
            mode,
            None,
        )?
        .ok_or_else(|| common::target_pruned(self.name()))?;
        Ok(Route {
            target: route_target,
            annotations: Default::default(),
            schema_version: Default::default(),
        })
    }
}

pub fn parse_synthesis_string(
    synthesis_string: &str,
    mode: AdaptMode,
) -> Result<common::PrecursorMap> {
    let parts: Vec<_> = synthesis_string
        .split(';')
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .collect();
    if parts.is_empty() {
        return Err(common::logic(
            "synllama",
            "adapter.route_string_empty",
            "empty synthesis string",
        ));
    }
    parse_synthesis(&parts, mode, "synllama")
}

fn parse_synthesis(
    parts: &[&str],
    mode: AdaptMode,
    adapter: &'static str,
) -> Result<common::PrecursorMap> {
    let template_indices: Vec<_> = parts
        .iter()
        .enumerate()
        .filter_map(|(index, part)| {
            part.strip_prefix('R')
                .filter(|suffix| {
                    !suffix.is_empty() && suffix.chars().all(|char| char.is_ascii_digit())
                })
                .map(|_| index)
        })
        .collect();
    let mut precursor_map = BTreeMap::new();
    let mut last_product: Option<String> = None;
    let mut reactant_start = 0;
    for template_index in template_indices {
        let product_index = template_index + 1;
        let Some(raw_product) = parts.get(product_index) else {
            return Err(common::logic(
                adapter,
                "adapter.route_string_invalid",
                format!("template {} has no product", parts[template_index]),
            ));
        };
        let product = match chem::normalize(raw_product) {
            Ok((smiles, _)) => smiles,
            Err(_) if mode == AdaptMode::Prune => {
                last_product = None;
                reactant_start = product_index + 1;
                continue;
            }
            Err(error) => return Err(error),
        };
        let mut reactants: Vec<_> = parts[reactant_start..template_index]
            .iter()
            .map(|reactant| (*reactant).to_owned())
            .collect();
        if let Some(previous_product) = last_product {
            reactants.push(previous_product);
        }
        if reactants.is_empty() {
            return Err(common::logic(
                adapter,
                "adapter.route_string_invalid",
                format!("no reactants found for product {raw_product}"),
            ));
        }
        precursor_map.insert(product.clone(), reactants);
        last_product = Some(product.to_string());
        reactant_start = product_index + 1;
    }
    Ok(precursor_map)
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::SynLlamaAdapter;
    use crate::{adapters::Adapter, route::AdaptMode};

    #[test]
    fn parses_ranked_template_syntax() {
        let route = SynLlamaAdapter
            .cast(
                json!({"synthesis_string": "C;CC;R1;CCO"}),
                AdaptMode::Strict,
                None,
            )
            .unwrap();
        assert_eq!(route.leaves().len(), 2);
    }

    #[test]
    fn pruning_does_not_carry_product_across_failed_step() {
        let route = SynLlamaAdapter
            .cast(
                json!({"synthesis_string": "C;R1;CC;C;R2;not-smiles;CCC;R3;CCO"}),
                AdaptMode::Prune,
                None,
            )
            .unwrap();
        assert_eq!(route.leaves()[0].value.smiles, "CCC");
    }
}

File: packages/retrocast-py/src/retrocast/adapters/synllama.py
from __future__ import annotations

from collections.abc import Iterator
from typing import Any

from pydantic import BaseModel, RootModel, ValidationError

from retrocast.adapters.base import AdaptMode, RawRouteEntry
from retrocast.adapters.common import build_molecule_from_precursor_map
from retrocast.adapters.errors import adapter_route_string_error, adapter_schema_error, adapter_target_mismatch
from retrocast.chem import canonicalize_smiles
from retrocast.exceptions import AdapterLogicError, InvalidSmilesError
from retrocast.models.route import Route
from retrocast.models.task import Target
from retrocast.typing import SmilesStr

# SECTION: Raw SynLlama Schema


class SynLlamaRouteInput(BaseModel):
    synthesis_string: str


class SynLlamaRouteList(RootModel[list[SynLlamaRouteInput]]):
    pass


# SECTION: Adapter


class SynLlamaAdapter:
    def iter_raw_routes(self, raw_payload: Any, *, source_key: str | None = None) -> Iterator[RawRouteEntry]:
        target_id = source_key or "<unknown>"
        try:
            routes = SynLlamaRouteList.model_validate(raw_payload)
        except ValidationError as exc:
            raise adapter_schema_error("synllama", target_id, "invalid route list") from exc
        for source_order, route in enumerate(routes.root, start=1):
            yield RawRouteEntry(payload=route, source_key=source_key, source_order=source_order)

    def cast(self, raw_route: Any, *, mode: AdaptMode = "strict", target: Target | None = None) -> Route:
        target_id = target.id if target is not None else "<unknown>"
        try:
            route = SynLlamaRouteInput.model_validate(raw_route)
        except ValidationError as exc:
            raise adapter_schema_error("synllama", target_id, "invalid route") from exc
        parts = [part.strip() for part in route.synthesis_string.split(";") if part.strip()]
        if not parts:
            raise adapter_route_string_error("synllama", "empty synthesis string", empty=True)
        try:
            parsed_target = canonicalize_smiles(parts[-1])
        except InvalidSmilesError:
            if mode == "prune":
                raise AdapterLogicError(
                    "SynLlama target molecule was pruned",
                    code="adapter.target_pruned",
                    context={"adapter": "synllama", "target_id": target_id},
                ) from None
            raise
        if target is not None:
            expected_smiles = canonicalize_smiles(target.smiles)
            if parsed_target != expected_smiles:
                raise adapter_target_mismatch(
                    "synllama", target.id, expected_smiles=expected_smiles, actual_smiles=parsed_target
                )
        precursor_map = self._parse_synthesis_string(route.synthesis_string, mode=mode)
        route_target = build_molecule_from_precursor_map(parsed_target, precursor_map, adapter="synllama", mode=mode)
        if route_target is None:
            raise AdapterLogicError(
                "SynLlama target molecule was pruned",
                code="adapter.target_pruned",
                context={"adapter": "synllama", "target_id": target_id},
            )
        return Route(target=route_target)

    def _parse_synthesis_string(self, synthesis_str: str, *, mode: AdaptMode = "strict") -> dict[SmilesStr, list[str]]:
        parts = [part.strip() for part in synthesis_str.split(";") if part.strip()]
        if not parts:
            raise adapter_route_string_error("synllama", "empty synthesis string", empty=True)

        template_indices = [index for index, part in enumerate(parts) if part.startswith("R") and part[1:].isdigit()]
        if not template_indices:
            return {}

        precursor_map: dict[SmilesStr, list[str]] = {}
        last_product_smiles: SmilesStr | None = None
        reactant_start = 0
        for template_index in template_indices:
            product_index = template_index + 1
            if product_index >= len(parts):
                raise adapter_route_string_error("synllama", "template has no product", fragment=parts[template_index])
            try:
                product_smiles = canonicalize_smiles(parts[product_index])
            except InvalidSmilesError:
                if mode == "strict":
                    raise
                last_product_smiles = None
                reactant_start = product_index + 1
                continue
            reactants = parts[reactant_start:template_index]
            if last_product_smiles is not None:
                reactants.append(last_product_smiles)
            if not reactants:
                raise adapter_route_string_error(
                    "synllama", "no reactants found for product", fragment=parts[product_index]
                )
            precursor_map[product_smiles] = reactants
            last_product_smiles = product_smiles
            reactant_start = product_index + 1
        return precursor_map
